README.rst
Transfluent API Client for Python
=================================

|build status|_

.. |build status| image:: https://secure.travis-ci.org/jpvanhal/python-transfluent.png?branch=master
   :alt: Build Status
.. _build status: http://travis-ci.org/jpvanhal/python-transfluent

This is a Python wrapper for the `Transfluent API <http://transfluent.com/backend-api/>`_.

Installation
------------

::

    $ pip install transfluent

Usage
-----

.. code-block:: python

    import transfluent

    # Initialize the Transfluent client and retrieve your authentication
    # token by using your email and password.
    client = transfluent.Transfluent()
    client.authenticate(email='example@example.org', password='my-password')

    # Alternatively, you may initialize the Transfluent client directly
    # with your authentication token.
    client = transfluent.Transfluent(token='my-token')

    # Order translations for a resource file.
    response = client.file_save(
        identifier='my-project/messages',
        language=1,
        file=open('translations/messages.pot'),
        type='po-file'
    )
    print "The file contains {0} words.".format(response['word_count'])
    response = client.file_translate(
        identifier='my-project/messages',
        language=1,
        target_languages=[11],
    )
    print "{0} words were ordered.".format(response['word_count'])

    # Check if the translation for the resource file is complete.
    is_translated = client.is_file_complete(
        identifier='my-project/messages',
        language=11
    )

    if is_translated:
        # Retrieve the translated resource file.
        content = client.file_read(
            identifier='my-project/messages',
            language=11
        )
        with open('translations/en/LC_MESSAGES/messages.po', 'w') as out:
            out.write(content)
    else:
        # Check the precise translation progress for the resource file.
        status = client.file_status(
            identifier='my-project/messages',
            language=11
       )
       print "Translation is {0} complete.".format(status['progress'])

Resources
---------

- `Issue Tracker <http://github.com/jpvanhal/python-transfluent/issues>`_
- `Code <http://github.com/jpvanhal/python-transfluent>`_
- `Development Version
  <http://github.com/jpvanhal/python-transfluent/zipball/master#egg=transfluent-dev>`_


CHANGES.rst
Changelog
---------

Here you can see the full list of changes between each Transfluent for Python
release.

0.3.0 (February 10, 2016)
^^^^^^^^^^^^^^^^^^^^^^^^^

- Added Python 3 support.

0.2.1 (May 29, 2013)
^^^^^^^^^^^^^^^^^^^^

- Fixed a bug in how `texts_translate` handled `target_languages` parameter.

0.2.0 (May 29, 2013)
^^^^^^^^^^^^^^^^^^^^

- Implemented Python wrappers for `Texts` and `TextsTranslate` methods in
  Transfluent's API

0.1.0 (April 22, 2013)
^^^^^^^^^^^^^^^^^^^^^^

- Initial public release.


test_transfluent.py
from io import BytesIO

from flexmock import flexmock
import pytest
import requests


def make_transfluent(*args, **kwargs):
    from transfluent import Transfluent
    return Transfluent(*args, **kwargs)


def make_transfluent_error(*args, **kwargs):
    from transfluent import TransfluentError
    return TransfluentError(*args, **kwargs)


def make_response(content, status_code=200):
    response = requests.Response()
    response.status_code = status_code
    response._content = content
    response.encoding = 'utf-8'
    return response


def make_error_response():
    content = (
        b'{"status":"ERROR","error":{"type":"EBackendParameterInvalid","mes'
        b'sage":"Name is required!"},"response":"Unfortunately an error occ'
        b'ured. You might want to try again. If problem persists, please re'
        b'port this as a bug. We are sorry for inconvenience."}'
    )
    return make_response(
        content=content,
        status_code=400
    )


class TestTransfluent(object):

    def test_constructor_without_parameters_sets_token_to_none(self):
        client = make_transfluent()
        assert client.token is None

    def test_constructor_sets_token(self):
        client = make_transfluent(token='foo')
        assert client.token == 'foo'

    def test_constructor_sets_transfluent_url(self):
        client = make_transfluent()
        assert client._transfluent_url == 'https://transfluent.com/v2/'

    def test_request_on_successful_json_response(self):
        response = make_response(b'{"status":"OK","response":"Hello World"}')
        (
            flexmock(requests)
            .should_receive('request')
            .with_args(
                'GET',
                'https://transfluent.com/v2/hello/World/',
                params=None
            )
            .and_return(response)
            .once()
        )
        client = make_transfluent()
        response = client._request('GET', 'hello/World/')
        assert response == u'Hello World'

    def test_request_on_successful_non_json_response(self):
        response = make_response(b'some content')
        (
            flexmock(requests)
            .should_receive('request')
            .with_args(
                'GET',
                'https://transfluent.com/v2/hello/World/',
                params=None
            )
            .and_return(response)
            .once()
        )
        client = make_transfluent()
        response = client._request('GET', 'hello/World/')
        assert response == b'some content'

    def test_request_on_error_raises_exception(self):
        from transfluent import TransfluentError
        response = make_error_response()
        (
            flexmock(requests)
            .should_receive('request')
            .with_args(
                'GET',
                'https://transfluent.com/v2/hello/',
                params=None
            )
            .and_return(response)
            .once()
        )
        client = make_transfluent()
        with pytest.raises(TransfluentError) as excinfo:
            response = client._request('GET', 'hello/')
        exception = excinfo.value
        assert exception.type == 'EBackendParameterInvalid'
        assert exception.message == 'Name is required!'

    def test_authed_request_without_parameters(self):
        client = make_transfluent(token='foo')
        fake_rv = flexmock()
        (
            flexmock(client)
            .should_receive('_request')
            .with_args('GET', 'customer/name/', {'token': 'foo'})
            .and_return(fake_rv)
            .once()
        )
        rv = client._authed_request('GET', 'customer/name/')
        assert rv is fake_rv

    def test_authed_request_with_parameters(self):
        client = make_transfluent(token='foo')
        fake_rv = flexmock()
        (
            flexmock(client)
            .should_receive('_request')
            .with_args(
                'POST',
                'customer/name/',
                {'token': 'foo', 'name': 'John'}
            )
            .and_return(fake_rv)
            .once()
        )
        rv = client._authed_request('POST', 'customer/name/', {'name': 'John'})
        assert rv is fake_rv

    def test_authenticate_sets_token(self):
        client = make_transfluent()
        (
            flexmock(client)
            .should_receive('_request')
            .with_args(
                'GET',
                'authenticate',
                {'email': 'john@example.com', 'password': 'test'}
            )
            .and_return({'token': 'foo'})
            .once()
        )
        client.authenticate(email='john@example.com', password='test')
        assert client.token == 'foo'

    def test_languages(self):
        client = make_transfluent()
        fake_rv = flexmock()
        (
            flexmock(client)
            .should_receive('_request')
            .with_args('GET', 'languages')
            .and_return(fake_rv)
            .once()
        )
        assert client.languages is fake_rv

    def test_retrieving_customer_name(self):
        client = make_transfluent()
        fake_rv = flexmock()
        (
            flexmock(client)
            .should_receive('_authed_request')
            .with_args('GET', 'customer/name')
            .and_return(fake_rv)
            .once()
        )
        assert client.customer_name is fake_rv

    def test_setting_customer_name(self):
        client = make_transfluent()
        (
            flexmock(client)
            .should_receive('_authed_request')
            .with_args('POST', 'customer/name', {'name': 'John Doe'})
            .once()
        )
        client.customer_name = 'John Doe'

    def test_retrieving_customer_email(self):
        client = make_transfluent()
        fake_rv = flexmock()
        (
            flexmock(client)
            .should_receive('_authed_request')
            .with_args('GET', 'customer/email')
            .and_return(fake_rv)
            .once()
        )
        assert client.customer_email is fake_rv

    def test_setting_customer_email(self):
        client = make_transfluent()
        (
            flexmock(client)
            .should_receive('_authed_request')
            .with_args(
                'POST',
                'customer/email',
                {'email': 'john.doe@example.com'}
            )
            .once()
        )
        client.customer_email = 'john.doe@example.com'

    def test_file_status(self):
        client = make_transfluent()
        fake_rv = flexmock()
        (
            flexmock(client)
            .should_receive('_authed_request')
            .with_args(
                'GET',
                'file/status',
                {'identifier': 'my-project/messages', 'language': 11}
            )
            .and_return(fake_rv)
            .once()
        )
        rv = client.file_status('my-project/messages', 11)
        assert rv is fake_rv

    def test_is_file_complete_when_file_has_been_translated(self):
        client = make_transfluent()
        (
            flexmock(client)
            .should_receive('file_status')
            .with_args('my-project/messages', 11)
            .and_return({'progress': '100%'})
            .once()
        )
        assert client.is_file_complete('my-project/messages', 11) is True

    def test_is_file_complete_when_file_has_not_been_translated(self):
        client = make_transfluent()
        (
            flexmock(client)
            .should_receive('file_status')
            .with_args('my-project/messages', 11)
            .and_return({'progress': '37.55%'})
            .once()
        )
        assert client.is_file_complete('my-project/messages', 11) is False

    def test_file_save_with_file_object(self):
        client = make_transfluent()
        fake_rv = flexmock()
        (
            flexmock(client)
            .should_receive('_authed_request')
            .with_args(
                'POST',
                'file/save',
                {
                    'identifier': 'my-project/messages',
                    'language': 1,
                    'format': 'UTF-8',
                    'content': 'ZmlsZSBjb250ZW50cw==',
                    'type': 'po-file',
                    'save_only_data': 0
                }
            )
            .and_return(fake_rv)
            .once()
        )
        rv = client.file_save(
            'my-project/messages',
            1,
            BytesIO(b'file contents'),
            'po-file'
        )
        assert rv is fake_rv

    def test_file_save_with_file_contents_as_string(self):
        client = make_transfluent()
        fake_rv = flexmock()
        (
            flexmock(client)
            .should_receive('_authed_request')
            .with_args(
                'POST',
                'file/save',
                {
                    'identifier': 'my-project/messages',
                    'language': 1,
                    'format': 'UTF-8',
                    'content': 'ZmlsZSBjb250ZW50cw==',
                    'type': 'po-file',
                    'save_only_data': 0
                }
            )
            .and_return(fake_rv)
            .once()
        )
        rv = client.file_save(
            'my-project/messages',
            1,
            "file contents",
            'po-file'
        )
        assert rv is fake_rv

    def file_translate(self):
        client = make_transfluent()
        fake_rv = flexmock()
        (
            flexmock(client)
            .should_receive('_authed_request')
            .with_args(
                'POST',
                'file/translate',
                {
                    'identifier': 'my-project/messages',
                    'language': 1,
                    'target_languages[]': [11, 14],
                    'comment': '',
                    'callback_url': '',
                    'level': 3
                }
            )
            .and_return(fake_rv)
            .once()
        )
        rv = client.file_translate('my-project/messages', 1, [11, 14])
        assert rv is fake_rv

    def test_file_read(self):
        client = make_transfluent()
        fake_rv = flexmock()
        (
            flexmock(client)
            .should_receive('_authed_request')
            .with_args(
                'GET',
                'file/read',
                {
                    'identifier': 'my-project/messages',
                    'language': 11,
                }
            )
            .and_return(fake_rv)
            .once()
        )
        rv = client.file_read('my-project/messages', 11)
        assert rv is fake_rv

    def test_texts_save(self):
        client = make_transfluent()
        fake_rv = flexmock()
        (
            flexmock(client)
            .should_receive('_authed_request')
            .with_args(
                'POST',
                'texts',
                {
                    'group_id': 'my-project/messages',
                    'language': 11,
                    'texts[foo]': 'bar',
                    'invalidate_translations': 1
                }
            )
            .and_return(fake_rv)
            .once()
        )
        rv = client.texts_save('my-project/messages', 11, {'foo': 'bar'})
        assert rv is fake_rv

    def test_texts_read(self):
        client = make_transfluent()
        fake_rv = flexmock()
        (
            flexmock(client)
            .should_receive('_authed_request')
            .with_args(
                'GET',
                'texts',
                {
                    'group_id': 'my-project/messages',
                    'language': 11,
                    'limit': 100,
                    'offset': 0
                }
            )
            .and_return(fake_rv)
            .once()
        )
        rv = client.texts_read('my-project/messages', 11)
        assert rv is fake_rv

    def test_texts_translate(self):
        client = make_transfluent()
        fake_rv = flexmock()
        (
            flexmock(client)
            .should_receive('_authed_request')
            .with_args(
                'GET',
                'texts/translate',
                {
                    'group_id': 'my-project/messages',
                    'source_language': 11,
                    'target_languages[]': [1, 14],
                    'texts[][id]': ['foo', 'bar'],
                    'comment': '',
                    'callback_url': '',
                    'max_words': 1000,
                    'level': 3
                }
            )
            .and_return(fake_rv)
            .once()
        )
        rv = client.texts_translate(
            group_id='my-project/messages',
            language=11,
            target_languages=[1, 14],
            texts=['foo', 'bar']
        )
        assert rv is fake_rv


class TestTransfluentError(object):
    def test_constructor_sets_response(self):
        response = make_error_response()
        exception = make_transfluent_error(response)
        assert exception.response is response

    def test_constructor_sets_type_from_response(self):
        response = make_error_response()
        exception = make_transfluent_error(response)
        assert exception.type == 'EBackendParameterInvalid'

    def test_constructor_sets_message_from_response(self):
        response = make_error_response()
        exception = make_transfluent_error(response)
        assert exception.message == 'Name is required!'

    def test_repr(self):
        response = make_error_response()
        exception = make_transfluent_error(response)
        assert (
            repr(exception) ==
            '<TransfluentError [EBackendParameterInvalid]>'
        )

    def test_str(self):
        response = make_error_response()
        exception = make_transfluent_error(response)
        assert str(exception) == 'Name is required!'


transfluent.py
# -*- coding: utf-8 -*-
"""
    transfluent
    ~~~~~~~~~~~

    :copyright: (c) 2013-2016 by Janne Vanhala.
    :license: BSD, see LICENSE for more details.
"""
import base64
import sys

import requests

__version__ = '0.3.0'

TRANSFLUENT_URL = 'https://transfluent.com/v2/'


PY2 = sys.version_info[0] == 2
if not PY2:
    iteritems = lambda x: iter(x.items())
else:
    iteritems = lambda x: x.iteritems()


class Transfluent(object):
    def __init__(self, token=None):
        self.token = token
        self._transfluent_url = TRANSFLUENT_URL

    def _request(self, method, path, data=None):
        url = self._transfluent_url + path
        kwargs = {}
        if method.upper() == 'GET':
            kwargs['params'] = data
        elif method.upper() == 'POST':
            kwargs['data'] = data
        else:
            raise ValueError('Unsupported request method: {0}'.format(method))
        response = requests.request(method, url, **kwargs)
        if response.status_code != 200:
            raise TransfluentError(response)
        try:
            data = response.json()
        except ValueError:
            return response.content
        else:
            return data['response']

    def _authed_request(self, method, path, data=None):
        data = data or {}
        data['token'] = self.token
        return self._request(method, path, data)

    def authenticate(self, email, password):
        data = {'email': email, 'password': password}
        response = self._request('GET', 'authenticate', data)
        self.token = response['token']

    @property
    def customer_name(self):
        return self._authed_request('GET', 'customer/name')

    @customer_name.setter
    def customer_name(self, name):
        self._authed_request('POST', 'customer/name', {'name': name})

    @property
    def customer_email(self):
        return self._authed_request('GET', 'customer/email')

    @customer_email.setter
    def customer_email(self, email):
        self._authed_request('POST', 'customer/email', {'email': email})

    @property
    def languages(self):
        return self._request('GET', 'languages')

    def texts_save(self, group_id, language, texts,
                   invalidate_translations=True):
        """
        Save texts to the system.

        :param group_id:
            Group id for texts (alphanumeric string, recommended max
            length 32 characters). All text ids in group are visible
            only in the group (e.g. text ids won't collide between
            groups).

        :type group_id: str

        :param language:
            The human language of the texts. E.g. 1 = English (GB),
            11 = Finnish. Please refer to :attr:`languages` for full
            list of language ids.

        :type language: int

        :param texts:
            A dict of text keys and content.

        :type texts: dict

        :param invalidate_translations:
            Optional. A boolean value which controls if translations are
            trashed or not. E.g. if you make a spelling error correction
            to the original text, you probably don't want to trash the
            translations and re-translate the text. Also if you save
            previously made translation to the system using this
            method, it's really important to set
            `invalidate_translations` off!

            Defaults to `True`.

            .. warning::
            Affects ALL texts set on the same request. If you need to
            invalidate some and keep others, you have to perform two
            requests.

        :type invalidate_translations: bool
        """
        data = {
            'group_id': group_id,
            'language': language,
            'invalidate_translations': 1 if invalidate_translations else 0,
        }
        for key, content in iteritems(texts):
            data['texts[{0}]'.format(key)] = content
        return self._authed_request('POST', 'texts', data)

    def texts_read(self, group_id, language, limit=100, offset=0):
        """
        Read texts from the system.

        :param group_id:
            Group id for texts (alphanumeric string, recommended max
            length 32 characters). All text ids in group are visible
            only in the group (e.g. text ids won't collide between
            groups).

        :type group_id: str

        :param language:
            The human language of the texts. E.g. 1 = English (GB),
            11 = Finnish. Please refer to :attr:`languages` for full
            list of language ids.

        :type language: int

        :type limit: int

        :type offset: int
        """
        data = {
            'group_id': group_id,
            'language': language,
            'limit': limit,
            'offset': offset,
        }
        return self._authed_request('GET', 'texts', data)

    def texts_translate(self, group_id, language, target_languages, texts,
                        **kwargs):
        """
        Order translations for new and changed texts.

        :param group_id:
            Group id for texts (alphanumeric string, recommended max
            length 32 characters). All text ids in group are visible
            only in the group (e.g. text ids won't collide between
            groups).

        :type group_id: str

        :param language:
            The human language the source text is written in.
            E.g. 1 = English (GB), 11 = Finnish. Please refer to
            :attr:`languages` for full list of language ids.

        :type language: int

        :param target_languages:
            A list of target languages to translate texts into. Please
            refer to :attr:`languages` for full list of language ids.

        :type target_languages: list

        :param texts:
            A list of text ids of texts to translate.

        :type texts: list

        :param callback_url:
            Optional. A GET request will be performed on this URL when
            the translation is ready.

        :type callback_url: str

        :param level:
            Optional. The level of tranlators. The supported values are:

            - `1`. A native speaker.
            - `2`. A professional translator.
            - `3`. Pair of translators.

            Defaults to `3`.

        :type level: int

        :param comment:
            Optional. Context information and other relevant information
            to the translator.

        :type comment: str

        :param max_words:
            Optional. To avoid surprisingly costly orders, you can
            specify maximum word count to order. If limit is exceeded,
            the order will fail. Defaults to `1000`.

        :type max_words: int
        """
        data = {
            'group_id': group_id,
            'source_language': language,
            'target_languages[]': target_languages,
            'texts[][id]': texts,
            'level': kwargs.get('level', 3),
            'comment': kwargs.get('comment', ''),
            'callback_url': kwargs.get('callback_url', ''),
            'max_words': kwargs.get('max_words', 1000)
        }
        return self._authed_request('GET', 'texts/translate', data)

    def file_save(self, identifier, language, file, type, format='UTF-8',
                  save_only_data=False):
        try:
            content = file.read()
        except AttributeError:
            content = file.encode(format)
        data = {
            'identifier': identifier,
            'language': language,
            'format': format,
            'content': base64.b64encode(content).decode('ascii'),
            'type': type,
            'save_only_data': int(save_only_data)
        }
        return self._authed_request('POST', 'file/save', data)

    def file_status(self, identifier, language):
        data = {
            'identifier': identifier,
            'language': language,
        }
        return self._authed_request('GET', 'file/status', data)

    def is_file_complete(self, identifier, language):
        status = self.file_status(identifier, language)
        return status['progress'] == '100%'

    def file_translate(self, identifier, language, target_languages, **kwargs):
        data = {
            'identifier': identifier,
            'language': language,
            'target_languages[]': target_languages,
            'level': kwargs.get('level', 3),
            'comment': kwargs.get('comment', ''),
            'callback_url': kwargs.get('callback_url', '')
        }
        return self._authed_request('POST', 'file/translate', data)

    def file_read(self, identifier, language):
        data = {
            'identifier': identifier,
            'language': language,
        }
        return self._authed_request('GET', 'file/read', data)


class TransfluentError(Exception):
    def __init__(self, response):
        data = response.json()
        self.response = response
        self.type = data['error']['type']
        self.message = data['error']['message']

    def __repr__(self):
        return '<TransfluentError [{0}]>'.format(self.type)

    def __str__(self):
        return self.message
